feat(comp/otelcol): component to filter OTEL telemetry

hainenber opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/agent/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description

#### Which issue(s) this PR fixes

<!-- Uncomment the following line if you want that GitHub issue gets automatically closed after merging the PR -->
Closes #4628 

#### Notes to the Reviewer

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [x] Config converters updated

ptodev commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,260 @@
+[river-strings]: {{< relref "../../config-language/expressions/types_and_values.md/#strings" >}}
+
+
+[OTTL]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/README.md
+[OTTL Converter functions]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/ottlfuncs#converters
+[HasAttrKeyOnDataPoint]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#hasattrkeyondatapoint
+[HasAttrOnDataPoint]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#hasattrondatapoint
+[OTTL booleans]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.85.0/pkg/ottl#booleans
+[OTTL math expressions]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.85.0/pkg/ottl#math-expressions
+[OTTL boolean expressions]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.85.0/pkg/ottl#boolean-expressions
+[OTTL resource context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlresource/README.md
+[OTTL scope context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlscope/README.md
+[OTTL span context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlspan/README.md
+[OTTL spanevent context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlspanevent/README.md
+[OTTL metric context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottlmetric/README.md
+[OTTL datapoint context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottldatapoint/README.md
+[OTTL log context]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.85.0/pkg/ottl/contexts/ottllog/README.md
```
Are all of those URLs used in this doc page?

hainenber replied:
Thanks for the review! I've removed the unused Markdown refs in latest commit.

ptodev commented on component/otelcol/processor/filter/filter.go:
```diff
@@ -0,0 +1,165 @@
+	Metric    []string `river:"metric,attr,optional"`
+	Datapoint []string `river:"datapoint,attr,optional"`
+}
+type logConfig struct {
```
Would you be able to also add the `include` and `exclude` properties please?

ptodev replied:
As we [agreed](https://github.com/grafana/agent/pull/5562#issuecomment-1782042834), there is no need to include these.

ptodev commented on component/otelcol/processor/filter/filter.go:
```diff
@@ -0,0 +1,165 @@
+	SpanEvent []string `river:"spanevent,attr,optional"`
+}
+
+type metricConfig struct {
```
Would you be able to also add the other config properties please:
* `include`
*  `exclude`
*  `regexp`

ptodev replied:
As we [agreed](https://github.com/grafana/agent/pull/5562#issuecomment-1782042834), there is no need to include these.

ptodev commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,260 @@
+{{% /admonition %}}
+
+You can specify multiple `otelcol.processor.filter` components by giving them different labels.
+
```
```suggestion

{{% admonition type="warning" %}}
Exercise caution when using `otelcol.processor.filter`:

- Make sure you understand the look of the incoming data and test the configuration thoroughly. 
  In general, use as specific a configuration as possible to lower the risk of the wrong data being dropped.
- [Orphaned Telemetry][]: The processor allows dropping spans. Dropping a span may lead to 
  orphaned spans if the dropped span is a parent. Dropping a span may lead to orphaned logs 
  if the log references the dropped span.

[Orphaned Telemetry]: https://github.com/open-telemetry/opentelemetry-collector/blob/v0.85.0/docs/standard-warnings.md#orphaned-telemetry
{{% /admonition %}}

```

ptodev reviewed (changes requested):
Thank you very much for this contribution! This has been a much-requested feature. I added a few comments for things we will need to change prior to merging.

hainenber commented on component/otelcol/processor/filter/filter.go:
```diff
@@ -0,0 +1,165 @@
+	})
+}
+
+type traceConfig struct {
```
Thanks for the review! Done in latest commit.

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+
+# otelcol.processor.filter
+
+{{< docs/shared lookup="flow/stability/experimental.md" source="agent" version="<AGENT VERSION>" >}}
```
```suggestion
{{< docs/shared lookup="flow/stability/experimental.md" source="agent" version="<AGENT_VERSION>" >}}
```

We are migrating to use an underscore in the `<AGENT_VERSION>` variable

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+* Resource attributes.
+* Instrumentation scope name.
+* Span attributes.
+If any condition is met, the telemetry data is dropped (each condition is ORed together)
```
```suggestion
If any condition is met, the telemetry data is dropped (each condition is ORed together).
```

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+* Span attributes.
+If any condition is met, the telemetry data is dropped (each condition is ORed together)
+
+In addition to the [standard OTTL Converter functions][OTTL Converter functions], 
+there is also a set of metrics-only functions exclusively by the processor
```
```suggestion
In addition to the [standard OTTL Converter functions][OTTL Converter functions], 
the following metrics-only functions are used exclusively by the processor:
```

I'm not sure this rephrase is accurate. The original looks like it's missing some words... and this is my best guess at what was intended.

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+{{% admonition type="warning" %}}
+Exercise caution when using `otelcol.processor.filter`:
+
+- Make sure you understand the look of the incoming data and test the configuration thoroughly. 
```
What is meant by "understand the look"?  The meaning isn't clear and I'm not sure what to suggest here.

hainenber replied:
It should be something akin to understanding the data's schema/format. Let me wordsmith this one

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+Exercise caution when using `otelcol.processor.filter`:
+
+- Make sure you understand the look of the incoming data and test the configuration thoroughly. 
+  In general, use as specific a configuration as possible to lower the risk of the wrong data being dropped.
```
```suggestion
  In general, use a configuration that is as specific as possible ensure you retain only the data you want to keep.
```

I've tried to rephrase this in a more active and positive way. Does this stay true to the intent of the concept described?

hainenber replied:
Very much so. Suggested change accepted. Thanks!

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+
+### traces block
+
+The `traces` block specifies statements which filter trace telemetry signals. 
```
```suggestion
The `traces` block specifies statements that filter trace telemetry signals. 
```

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+1. `metric`
+2. `datapoint`
+
+Hence, if at least one `metric` condition is satisfied, the `datapoint` conditions will not be checked.
```
```suggestion
If at least one `metric` condition is satisfied, the `datapoint` conditions will not be checked.
```

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+
+### metrics block
+
+The `metrics` block specifies statements which filter metric telemetry signals. 
```
```suggestion
The `metrics` block specifies statements that filter metric telemetry signals. 
```

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+1. `span`
+2. `spanevent`
+
+Hence, if at least one `span` condition is satisfied, the `spanevent` conditions will not be checked.
```
```suggestion
If at least one `span` condition is satisfied, the `spanevent` conditions will not be checked.
```

clayton-cornell commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+
+### logs block
+
+The `logs` block specifies statements which filter log telemetry signals. 
```
```suggestion
The `logs` block specifies statements that filter log telemetry signals. 
```

clayton-cornell reviewed (commented):
Some doc suggestions and questions

ptodev commented:
> Appreciate your review effort @ptodev! Other comments totally makes sense to me but there are three with similar concept so I'd like to discuss jointly.
> 
> I saw that the config schema with `include`, `exclude` and alike are at brink of being deprecated, as dictated in the [ref](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#alternative-config-options). Should we also implement it for a brand new Flow component? I feel like it could be technical baggage and we'll revisit for the seemingly inevitable removal.

@hainenber Thank you for pointing this out! I did not realise it. In that case I agree that we can just have the OTTL parameters. We don't need to include these other ones which will be deprecated.

ptodev commented on component/otelcol/processor/filter/filter.go:
```diff
@@ -0,0 +1,165 @@
+	LogRecord []string `river:"log_record,attr,optional"`
+}
+
+type Arguments struct {
```
As we [agreed](https://github.com/grafana/agent/pull/5562#issuecomment-1782042834), there is no need to include these.

ptodev commented:
@hainenber Would you mind addressing @clayton-cornell's comments please, and rebasing the branch? I will then take a final look and will likely merge it. Thanks again very much for your work on this component!

hainenber commented:
> @hainenber Would you mind addressing @clayton-cornell's comments please, and rebasing the branch? I will then take a final look and will likely merge it. Thanks again very much for your work on this component!

@ptodev I've addressed Clayton's comments and rebased the branch. PTAL when you're back. Thanks!

ptodev commented on docs/sources/flow/reference/components/otelcol.processor.filter.md:
```diff
@@ -0,0 +1,302 @@
+  * `end_time_unix_nano - start_time_unix_nano`
+  * `sum([1, 2, 3, 4]) + (10 / 1) - 1`
+
+{{% admonition type="note" %}}
+Some characters inside River strings [need to be escaped][river-strings] with a `\` character.
+For example, the OTTL statement `attributes["grpc"] == true` 
+is written in River as `attributes[\"grpc\"] == true`.
+
+[river-strings]: {{< relref "../../config-language/expressions/types_and_values.md/#strings" >}}
+{{% /admonition %}}
```
The next release of the agent will support raw strings, so I updated some of the documentation in #5636. I'm happy to update the `otelcol.processor.filter` docs after your PR is merged though. There's no need to mention raw strings in your PR.

ptodev reviewed (approved):
Thank you, it looks good! I left a few minor comments, but the only thing which really needs to be fixed is just the non-working unit test.

hainenber commented:
Thank you for another review! I've accepted most of your suggested changes and moved the `convert()` to `types.go` file. PTAL if it's OK by now.
